src/library_manager/lib_manager.c

```diff
@@ -509,98 +519,98 @@ static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext,
 	if (ret < 0)
 		goto return_on_fail;
 
-	dcache_invalidate_region((void *)dma_ext->dmabp.addr, size);
+	dcache_invalidate_region((void __sparse_cache *)dma_ext->dmabp.addr, size);
 
 return_on_fail:
 	rfree(dma_block_cfg);
 	return ret;
 }
 
-static int lib_manager_store_data(struct lib_manager_dma_ext *dma_ext, void *dst_addr,
-				  uint32_t dst_size)
+static int lib_manager_store_data(struct lib_manager_dma_ext *dma_ext,
+				  void __sparse_cache *dst_addr, uint32_t dst_size)
 {
+	struct lib_manager_dma_buf *const dma_buf = &dma_ext->dmabp;
 	uint32_t copied_bytes = 0;
-	uint32_t bytes_to_copy = 0;
-	struct lib_manager_dma_buf *dma_buf = &dma_ext->dmabp;
-	int ret;
 
 	while (copied_bytes < dst_size) {
+		uint32_t bytes_to_copy;
+		int ret;
+
 		if ((dst_size - copied_bytes) > MAN_MAX_SIZE_V1_8)
 			bytes_to_copy = MAN_MAX_SIZE_V1_8;
 		else
-			bytes_to_copy = (dst_size - copied_bytes);
+			bytes_to_copy = dst_size - copied_bytes;
 
-		lib_manager_dma_buffer_update(dma_buf, dma_ext->dmabp.addr,
-					      bytes_to_copy);
+		lib_manager_dma_buffer_update(dma_buf, bytes_to_copy);
 
 		ret = lib_manager_load_data_from_host(dma_ext, bytes_to_copy);
 		if (ret < 0)
 			return ret;
-		memcpy_s((void *)((char *)dst_addr + copied_bytes), bytes_to_copy,
-			 (void *)dma_ext->dmabp.addr, bytes_to_copy);
+		memcpy_s((__sparse_force uint8_t *)dst_addr + copied_bytes, bytes_to_copy,
+			 (void *)dma_buf->addr, bytes_to_copy);
 		copied_bytes += bytes_to_copy;
 	}
+
 	dcache_writeback_region(dst_addr, dst_size);
 
 	return 0;
 }
 
-static int lib_manager_allocate_store_mem(uint32_t *address, uint32_t size, uint32_t attribs)
+static void __sparse_cache *lib_manager_allocate_store_mem(uint32_t size,
+							   uint32_t attribs)
 {
-	void *local_add;
+	void __sparse_cache *local_add;
 #if CONFIG_L3_HEAP
 	uint32_t caps = SOF_MEM_CAPS_L3 | SOF_MEM_CAPS_DMA;
 
-	/* allocate new buffer */
-	local_add = rmalloc(SOF_MEM_ZONE_SYS, 0, caps, size);
+	/* allocate new buffer: cached alias */
+	local_add = (__sparse_force void __sparse_cache *)rmalloc(SOF_MEM_ZONE_SYS, 0, caps, size);
 #else
 	uint32_t addr_align = CONFIG_MM_DRV_PAGE_SIZE;
 	uint32_t caps = SOF_MEM_CAPS_DMA;
 
-	/* allocate new buffer */
-	local_add = rballoc_align(0, caps, size, addr_align);
+	/* allocate new buffer: cached alias */
+	local_add = (__sparse_force void __sparse_cache *)rballoc_align(0, caps, size, addr_align);
 #endif
 	if (!local_add) {
 		tr_err(&lib_manager_tr, "dma_buffer_init(): alloc failed");
-		return -ENOMEM;
+		return NULL;
```

Comment:
You changed the function declarations to return void, so why are you returning a value?

Reply:
@softwarecki it's returning `void *`

Reply:
my mistake, sorry